db/version_builder.cc

```diff
@@ -485,12 +485,12 @@ class VersionBuilder::Rep {
     if (add_it != add_files.end()) {
       UnrefFile(add_it->second);
       add_files.erase(add_it);
-    } else {
-      auto& del_files = level_state.deleted_files;
-      assert(del_files.find(file_number) == del_files.end());
-      del_files.emplace(file_number);
     }
 
+    auto& del_files = level_state.deleted_files;
+    assert(del_files.find(file_number) == del_files.end());
+    del_files.emplace(file_number);
```

Comment:
I'm not sure we need to keep track of the deletion in the case of addition followed by deletion of the same file (although I also don't know if that case can happen).

Reply:
During recovery, when we replay the manifest, all edits for a CF go through the same `VersionBuilder`, so it is possible for a single builder to see the deletion of a file that wasn't in the base `Version` but got added in an earlier `VersionEdit`.

Regarding the change here, I think the addition and deletion code have to be symmetrical, otherwise one can end up with incorrect results. As an example, let's assume file X is initially on level Y. If we make the change in `ApplyFileAddition` but not here, then after `del(levelX, fileY) -> add(levelX, fileY) -> del(levelX, fileY)` the file would be still alive due to it being present in the base version, which is not correct.

Reply:
Understood now, thanks.